open_spiel/games/sheriff.cc

```diff
@@ -0,0 +1,392 @@
+// Copyright 2020 DeepMind Technologies Ltd.All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.#include
+// "third_party/open_spiel/games/battleship.h"
+
+#include "open_spiel/games/sheriff.h"
+
+#include <algorithm>
+
+#include "open_spiel/abseil-cpp/absl/strings/str_cat.h"
+
+namespace open_spiel {
+namespace sheriff {
+namespace {
+inline constexpr const Player kSMUGGLER = Player{0};
+inline constexpr const Player kSHERIFF = Player{1};
+
+template <typename T>
+void StrAppendVector(std::string* s, const std::vector<T>& v) {
+  absl::StrAppend(s, "[");
+  for (size_t index = 0; index < v.size(); ++index) {
+    if (index > 0) absl::StrAppend(s, ",");
+    absl::StrAppend(s, v[index]);
+  }
+  absl::StrAppend(s, "]");
+}
+}  // namespace
+
+SheriffState::SheriffState(
+    const std::shared_ptr<const SheriffGame> sheriff_game)
+    : State(sheriff_game), sheriff_game_(sheriff_game) {}
+
+Player SheriffState::CurrentPlayer() const {
+  if (!num_illegal_items_) {
+    // The smuggler still hasn't decided the number of illegal items to
+    // place in the cargo. The game has just begun.
+    return kSMUGGLER;
+  } else if (bribes_.size() == inspection_feedback_.size()) {
+    // The smuggler has received feedback for all the bribes.
+    //
+    // If the number of bribes is equal to the number of bribing turns in the
+    // game, the game is over.
+    if (bribes_.size() == sheriff_game_->conf.num_rounds) {
+      return kTerminalPlayerId;
+    } else {
+      // Otherwise, a new bribing round begins.
+      return kSMUGGLER;
+    }
+  } else {
+    // The smuggles has made a bribe, but no feedback has been given out yet.
+    return kSHERIFF;
+  }
+}
+
+std::vector<Action> SheriffState::LegalActions() const {
+  const SheriffGame::SheriffGameConfiguration& conf = sheriff_game_->conf;
+
+  std::vector<Action> action_ids;
+  if (IsTerminal()) {
+    return {};
+  } else if (!num_illegal_items_) {
+    // This is the beginning of the game. The smuggles must decide how many
+    // illegal items to place in the cargo, which must be an integer in the
+    // range [0, conf.max_items]. The action id will correspond to the number
+    // of illegal items placed in the cargo.
+
+    action_ids.reserve(conf.max_items + 1);
+    for (uint32_t num_illegal_items = 0; num_illegal_items <= conf.max_items;
```

Comment:
I'm fine with `uint32`, it's just I didn't see it used very much in the library, as usually `int`s are used for storing the settings. I guess it's ok according to the [style guide](https://google.github.io/styleguide/cppguide.html), which just specifies to not use `unsigned` qualifier.

Reply:
That's a good point! Let's see what the maintainers prefer, I don't mind moving to `int` and add checks for non-negativity where needed.

Reply:
I think uint32 is ok, thanks.